src/main/scala/com/softwaremill/realworld/users/UsersService.scala

```diff
@@ -46,27 +46,48 @@ class UsersService(authService: AuthService, usersRepository: UsersRepository):
     val passwordClean = user.password.trim()
 
     for {
+      _ <- validateEmail(emailClean)
       maybeUser <- usersRepository.findUserWithPasswordByEmail(emailClean)
       userWithPassword <- ZIO.fromOption(maybeUser).mapError(_ => Unauthorized())
       _ <- authService.verifyPassword(passwordClean, userWithPassword.hashedPassword)
       jwt <- authService.generateJwt(emailClean)
     } yield userWithPassword.user.copy(token = Some(jwt))
   }
 
-  def update(updateData: UserUpdateData, userId: Int): IO[Throwable, User] = for {
-    oldUser <- usersRepository
-      .findUserWithPasswordById(userId)
-      .someOrFail(NotFound(UserWithIdNotFoundMessage(userId)))
-    password <- updateData.password
-      .map(newPassword => authService.encryptPassword(newPassword))
-      .getOrElse(ZIO.succeed(oldUser.hashedPassword))
-    updatedUser <- usersRepository
-      .updateById(
-        updateData.update(oldUser.copy(hashedPassword = password)),
-        userId
-      )
-      .someOrFail(NotFound(UserWithIdNotFoundMessage(userId)))
-  } yield updatedUser
+  def update(updateData: UserUpdateData, userId: Int): IO[Throwable, User] =
+    val emailCleanOpt = updateData.email.map(email => email.toLowerCase.trim())
+    val usernameCleanOpt = updateData.username.map(username => username.trim())
+
+    def checkUserDoesNotExist(emailOpt: Option[String], usernameOpt: Option[String]): IO[Exception, Unit] =
+      (emailOpt, usernameOpt) match
+        case (Some(email), None)    => checkUserDoesNotExistByEmail(email)
+        case (None, Some(username)) => checkUserDoesNotExistByUsername(username)
+        case (Some(email), Some(username)) =>
+          checkUserDoesNotExistByEmail(email)
+          checkUserDoesNotExistByUsername(username)
+        case (None, None) => ZIO.unit
+
+    def validateEmailOpt(emailCleanOpt: Option[String]): IO[Exception, Unit] =
+      emailCleanOpt match
```

Comment:
**suggestion:** This is can be expressed with `forEach` (which is an equivalent of `traverse`)
```scala
ZIO.forEach(emailCleanOpt)(validateEmail).unit
```

btw check out https://zio.dev/reference/control-flow/